core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -287,7 +262,81 @@ public void messageReceived(final Request request, final TransportChannel channe
 
         @Override
         public void messageReceived(Request request, TransportChannel channel, Task task) throws Exception {
-            new PrimaryPhase((ReplicationTask) task, request, channel).run();
+            ReplicationTask replicationTask = (ReplicationTask) task;
+            boolean success = false;
+            PrimaryShardReference primaryShardReference = getPrimaryShardReference(request.shardId());
+            try {
+                if (primaryShardReference.isRelocated()) {
+                    setPhase(replicationTask, "primary_delegation");
+                    // delegate primary phase to relocation target
+                    // it is safe to execute primary phase on relocation target as there are no more in-flight operations where primary
+                    // phase is executed on local shard and all subsequent operations are executed on relocation target as primary phase.
+                    final ShardRouting primary = primaryShardReference.routingEntry();
+                    assert primary.relocating() : "indexShard is marked as relocated but routing isn't" + primary;
+                    DiscoveryNode relocatingNode = clusterService.state().nodes().get(primary.relocatingNodeId());
+                    transportService.sendRequest(relocatingNode, transportPrimaryAction, request, transportOptions,
+                        new TransportChannelResponseHandler<Response>(logger, channel, "rerouting indexing to target primary " + primary) {
+                            @Override
+                            public Response newInstance() {
+                                return newResponseInstance();
+                            }
+
+                            @Override
+                            protected void onFinished() {
+                                setPhase(replicationTask, "finished");
+                            }
+                        });
+                } else {
+                    setPhase(replicationTask, "primary");
+                    final IndexMetaData indexMetaData = clusterService.state().getMetaData().index(request.shardId().getIndex());
+                    final boolean executeOnReplicas = (indexMetaData == null) || shouldExecuteReplication(indexMetaData.getSettings());
```

Comment:
Deciding on whether to replicate to shadow indices is dependent on the action in hand. For example, flush does go to all shards. This is why I made it a boolean parameter, with the idea to not have to go and touch the various actions too much.